crypto/benches/mls_proteus.rs

```diff
@@ -0,0 +1,268 @@
+        }
+    }
+    group.finish();
+}
+
+fn update_client_bench(c: &mut Criterion) {
+    let mut group = c.benchmark_group("Mls vs Proteus: update");
+    for i in (GROUP_RANGE).step_by(GROUP_STEP) {
+        // MLS
+        for (ciphersuite, credential, in_memory, bench_name) in mls_cases() {
+            group.bench_with_input(BenchmarkId::new(bench_name, i), &i, |b, i| {
+                b.to_async(FuturesExecutor).iter_batched(
+                    || {
+                        let (mut central, id) = setup_mls(ciphersuite, &credential, in_memory);
+                        add_clients(&mut central, &id, ciphersuite, *i);
+                        (central, id)
+                    },
+                    |(mut central, id)| async move {
+                        black_box(central.update_keying_material(&id).await.unwrap());
+                        black_box(central.commit_accepted(&id).await.unwrap());
+                    },
+                    BatchSize::SmallInput,
+                )
+            });
+        }
+        // Proteus
+        // From proteus POV, adding 1 client in a group of N means adding N times 1 client to N central
+        // To simplify we are just going to add N times a client to just 1 central
+        for (in_memory, bench_name) in proteus_cases() {
+            group.bench_with_input(BenchmarkId::new(bench_name, i), &i, |b, i| {
+                b.to_async(FuturesExecutor).iter_batched(
+                    || {
+                        let (mut central, keystore) = setup_proteus(in_memory);
+                        let session_material = (0..*i)
+                            .map(|_| (session_id(), new_prekey().serialise().unwrap()))
+                            .collect::<Vec<(String, Vec<u8>)>>();
+                        for (session_id, key) in &session_material {
+                            block_on(async {
+                                central.session_from_prekey(session_id, key).await.unwrap();
+                                central.session_save(&keystore, session_id).await.unwrap();
+                            });
+                        }
+                        (central, keystore, session_material)
+                    },
+                    |(mut central, keystore, session_material)| async move {
+                        for (session_id, _) in session_material {
+                            // generate a new prekey
+                            let pkb = black_box(
+                                PreKeyBundle::new(
+                                    keys::IdentityKeyPair::new().public_key,
+                                    &PreKey::new(keys::PreKeyId::new(2)),
+                                )
+                                .serialise()
+                                .unwrap(),
+                            );
+                            // replace existing session
+                            black_box(central.session_from_prekey(&session_id, &pkb).await.unwrap());
```

Comment:
Maybe more atomic would be the use case we're looking for?
```suggestion
                            black_box(central.session_delete(&keystore, &session_id).await.unwrap());
                            black_box(central.session_from_prekey(&session_id, &pkb).await.unwrap());
```